Gsdx: Add Mipmap (Software mode) and preload data frame to GUI.

ssakash opened the pull request:
~~Windows build for testing : https://www.dropbox.com/s/o4xb7n2z1fo6a1s/GSdx32-SSE2.dll?dl=0~~ outdated build of the commit.

willkuer commented:
This is only for troubleshooting, or? Does mipmapping has negative effects?

ssakash commented:
I haven't observed any flaws of Mipmapping on software mode. the intent of the code is for easier debugging and to also keep up consistency with Linux GUI which also has option for mipmap.

gregory38 commented on plugins/GSdx/GSRenderer.cpp:
```diff
@@ -574,11 +574,13 @@ void GSRenderer::KeyEvent(GSKeyEventData* e)
 		case VK_DELETE:
 			m_aa1 = !m_aa1;
-			printf("GSdx: (Software) aa1 is now %s.\n", m_aa1 ? "enabled" : "disabled");
+			theApp.SetConfig("aa1", m_aa1);
```
Wasn't hotkey change suppose to be temporary?

ssakash replied:
usually, but I don't think there are any benefits in a temporary change. users couldn't find whether currently Mipmap is enabled (or) not after using multiple toggles without the following change. ( for those who don't use the console log). This change is currently important due to fact that users might take a look at the GUI for current status of Mipmap and AA1, outdated value might confuse them.

gregory38 commented on plugins/GSdx/GSSetting.cpp:
```diff
@@ -124,10 +124,10 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Enables mipmapping, which some games require to render correctly. Turn off only for debug purposes.";
-#endif
+		case IDC_PRELOAD_GS:
+			return "The Following Hack uploads GS data when rendering a new frame to reprdouce some effects accurately, Fixes Black Screen issues in games like Primal. ";
```
Search in the issue report the other game that require the option.

ssakash replied:
Added Armored core: Last Raven to the tool tip instead of Primal, since Primal was only fixed on replay of dump while using ``Preload Data Frame`` not while playing the game itself.

gregory38 commented on plugins/GSdx/GSSettingsDlg.cpp:
```diff
@@ -423,6 +430,7 @@ void GSSettingsDlg::UpdateControls()
 		EnableWindow(GetDlgItem(m_hWnd, IDC_UPSCALE_MULTIPLIER_TEXT), hw);
 		EnableWindow(GetDlgItem(m_hWnd, IDC_FILTER), hw);
 		EnableWindow(GetDlgItem(m_hWnd, IDC_PALTEX), hw);
+		EnableWindow(GetDlgItem(m_hWnd, IDC_PRELOAD_GS), hw);
```
For me it must go into the hack section.

ssakash replied:
I'll move it to hacks dialog , my fault. :p

ssakash commented:
**Review of updated change with respect to master branch**

* added option for preload data frame to hacks dialog.

* made mipmap and AA1 hotkey to do a permanent change similar to shader toggle hotkey for preventing any GUI conflicts.

*added option for enabling/disabling mipmap on software mode options groupbox.

turtleli commented on plugins/GSdx/GSSetting.cpp:
```diff
@@ -124,10 +124,10 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Enables mipmapping, which some games require to render correctly. Turn off only for debug purposes.";
-#endif
+		case IDC_PRELOAD_GS:
+			return "The Following Hack uploads GS data when rendering a new frame to reprdouce some effects accurately, Fixes Black Screen issues in games like Armored Core: Last Raven. ";
```
Either make IDC_PRELOAD_GS Windows only, or add an enum for Linux, otherwise it'll break Linux compilation.

The tooltip string needs fixed. "The Following Hack" isn't needed, reproduce is misspelt, comma should be full stop, black screen does not need capitals, and there's a space at the end.

ssakash replied:
added it to the previous #ifdef  :heavy_check_mark:

turtleli commented on plugins/GSdx/GSdx.rc:
```diff
@@ -213,31 +214,32 @@ BEGIN
     COMBOBOX        IDC_CRC_LEVEL,105,206,127,63,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
     CONTROL         "Enable HW Hacks",IDC_HACKS_ENABLED,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,11,223,71,10
-    PUSHBUTTON      "Configure Hacks...",IDC_HACKSBUTTON,105,221,127,14
+    PUSHBUTTON      "Configure Hacks",IDC_HACKSBUTTON,105,221,127,14
```
This makes it inconsistent with the "Configure Shade Boost..." button. Either undo this change or change the shade boost button as well.

Also, might be a good idea to provide screenshots of the dialogs.

ssakash replied:
I remove the ... stuffs since it's just a waste of space, can't really post the screenshots right now since I did these changes on Git online editor through my mobile. would be nice if you could double check the recent changes :)

gregory38 commented:
We really need to clarify the permanent/temporary option when it is changed by hotkey.
@bositman, please give your opinion on the 2nd commit.Thanks

bositman commented:
I think the hotkey change should not be saved in the ini. It's an option that rarely gets toggled and when it does it's almost always to debug sutff / check for issues. Thus keep the hotkey as a temporary toggle and not permanent

ssakash commented:
modified second commit to only deal with the console log message changes. (removed the permanent change stuffs) 

but in my humble opinion, temporary change doesn't give any benefits over permanent  change. the former only misleads users due to the outdated GUI status of the options.

turtleli commented on plugins/GSdx/GSSetting.cpp:
```diff
@@ -123,11 +123,12 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
+		// Exclusive for Hardware Renderer	
+		case IDC_PRELOAD_GS:
+			return " Uploads GS data when rendering a new frame to reproduce some effects accurately. Fixes Black Screen issues in games like Primal.";
```
Black Screen -> black screen

turtleli commented on plugins/GSdx/GSRenderer.cpp:
```diff
@@ -574,11 +574,11 @@ void GSRenderer::KeyEvent(GSKeyEventData* e)
 		case VK_DELETE:
 			m_aa1 = !m_aa1;
-			printf("GSdx: (Software) aa1 is now %s.\n", m_aa1 ? "enabled" : "disabled");
+			printf("GSdx: (Software) Edge Anti-aliasing is now %s.\n", m_aa1 ? "enabled" : "disabled");
```
"Anti" IMO should be "anti"

You should probably make the Linux and Windows messages the same as well (including whether they are printed to stdout or stderr)

turtleli commented:
Maybe the best thing to do is to provide an option on whether the toggle is temporary or permanent. That would satisfy everyone - or not, because then people would argue about what the default behaviour should be.

ssakash commented:
> Maybe the best thing to do is to provide an option on whether the toggle is temporary or permanent. That would satisfy everyone - **or not, because then people would argue about what the default behaviour should be**.

you got the point. I might consider opening an issue/poll later for knowing user preference on either temporary / permanent change.

Fixed those other concerns on the toggle messages and uppercase -> lowercase stuffs. Thanks

turtleli commented on plugins/GSdx/GSRenderer.cpp:
```diff
@@ -610,19 +610,19 @@ void GSRenderer::KeyEvent(GSKeyEventData* e)
 		case XK_Prior:
 			m_fxaa = !m_fxaa;
-			fprintf(stderr,"GSdx: fxaa is now %s.\n", m_fxaa ? "enabled" : "disabled");
+			printf("GSdx: fxaa is now %s.\n", m_fxaa ? "enabled" : "disabled");
```
Use the same message as Windows here?

turtleli commented on plugins/GSdx/GSSetting.cpp:
```diff
@@ -123,11 +123,12 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
+		// Exclusive for Hardware Renderer	
+		case IDC_PRELOAD_GS:
+			return " Uploads GS data when rendering a new frame to reproduce some effects accurately. Fixes black Screen issues in games like Primal.";
```
Screen -> screen

ssakash commented:
- changed Linux FXAA toggle message to be similar to windows one.
- Screen -> screen

turtleli commented on plugins/GSdx/GSdx.rc:
```diff
@@ -89,13 +89,14 @@ IDB_LOGOGL              BITMAP                  "res\\logo-ogl.bmp"
-    GROUPBOX        "USE AT YOUR OWN RISK!",IDC_STATIC,7,7,133,155,0,WS_EX_TRANSPARENT
+    DEFPUSHBUTTON   "OK",IDOK,84,181,56,14
+    GROUPBOX        "USE AT YOUR OWN RISK!",IDC_STATIC,7,7,133,171,0,WS_EX_TRANSPARENT
+	CONTROL         "Preload Data Frame",IDC_PRELOAD_GS,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,14,95,80,8
```
Use spaces instead of a tab at the start of this line.

Also can you move the "Software Mode Settings" Groupbox in the main dialog up 2 dialog units so that the spacing is more consistent.

I can merge after that's done.

ssakash replied:
done :+1: 

I hope you meant to decrease 2 on the y co-ordinate ? coz that's what I thought.

turtleli commented on plugins/GSdx/GSSetting.cpp:
```diff
@@ -123,11 +123,12 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
+		// Exclusive for Hardware Renderer	
+		case IDC_PRELOAD_GS:
+			return " Uploads GS data when rendering a new frame to reproduce some effects accurately. Fixes black screen issues in games like Primal.";
```
Oops, I missed this one. Remove the space before Uploads.

Then I'll really merge it.

ssakash replied:
fixed

turtleli commented on plugins/GSdx/GSSetting.cpp:
```diff
@@ -123,11 +123,12 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
+		// Exclusive for Hardware Renderer	
+		case IDC_PRELOAD_GS:
+			return "Uploads GS data when rendering a new frame to reproduce some effects accurately. Fixes black screen issues in games like Primal.";
```
..It must seem like I'm just messing around with you. I assure you I am not.

Wasn't it supposed to say Armored Core: Last Raven? Oh, also do a trailing whitespace check.

ssakash replied:
> Wasn't it supposed to say Armored Core: Last Raven? Oh, also do a trailing whitespace check.

yeah, you're right. no clue how that got swapped up @_@ 

> ..It must seem like I'm just messing around with you. I assure you I am not.

I'm not sure what gave you that impression ! 

either way, fixed the trailing whitespace and ``IDC_PRELOAD_GS`` tooltip modification.

turtleli replied:
Well, most people don't find a few issues each time. Bad habit of mine :)

ssakash commented on plugins/GSdx/GSSetting.cpp:
```diff
@@ -123,11 +123,12 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Makes textures partially or fully transparent as required by emulation. May cause unusual slowdowns for some games.";
 		case IDC_LOGZ:
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
+		// Exclusive for Hardware Renderer
```
I'm not sure why I bothered to even add this useless comment ? -_-
